Bluetooth: Classic: OBEX enablement

lylezhu2012 opened the pull request:
* Bluetooth: GOEP: Enable GOEP feature

Add a Kconfig BT_GOEP to control the GOEP feature.

Implement the GOEP protocol and transport, both for GOEP 1.1 and GOEP 2.x.

For GOEP transport, OBEX over RFCOMM, and OBEX over L2CAP are supported.

For GOEP protocol, `put`, `get`, `abort`, `setpath`, and `action` are supported. And only one operation can be processed at the same time. The feature `Reliable Session` is unsupported.

* Bluetooth: GOEP: Improve the MTU configuration

The MTU of the GOEP should be not less than 255. So, if the transport is RFCOMM, the CONFIG_BT_BUF_ACL_RX_SIZE should be not less than 264.
It includes,
- 255 bytes for the minimum MTU of GOEP,
- 4 bytes for L2CAP Header,
- 5 bytes for RFCOMM header and FCS.

And if the transport is L2CAP, the CONFIG_BT_BUF_ACL_RX_SIZE should be not less than 259.
It includes,
- 255 bytes for the minimum MTU of GOEP,
- 4 bytes for L2CAP Header.

Add Kconfig `BT_GOEP_RFCOMM_MTU` to configure the maximum size for RFCOMM transport. The range of `BT_GOEP_RFCOMM_MTU` is `[264, BT_RFCOMM_L2CAP_MTU]`. And the GOEP MTU via RFCOMM transport should be in the range `[255, (BT_GOEP_RFCOMM_MTU-9)]`.

Add Kconfig `BT_GOEP_L2CAP_MTU` to configure the maximum size for L2CAP transport. The range of `BT_GOEP_L2CAP_MTU` is `[259, BT_BUF_ACL_RX_SIZE]`. And the GOEP MTU via L2CAP transport should be in the range `[255, (BT_GOEP_L2CAP_MTU-4)]`.

* Bluetooth: shell: Add transport commands for GOEP

Add commands for GOEP to test transport features.

Add commands for transport over RFCOMM, including `register-rfcomm`, `connect-rfcomm`, and `disconnect-rfcomm`.

Add commands for transport over L2CAP, including `register-l2cap`, `connect-l2cap`, and `disconnect-l2cap`.

* Bluetooth: shell: Add OBEX commands for GOEP

Add commands for GOEP to test OBEX features.

Add command `alloc-buf` and `release-buf` to allocate and release TX buffer.

Add command set `add-header` to add the OBEX headers to allocated TX buffer.

Add command set `client` to send OBEX client requests with allocated TX buffer.

Add command set `server` to send OBEX responses with allocated TX buffer.

* Bluetooth: OBEX: Check if the added string is valid

There are types of string can be added by using OBEX adding header function bt_obex_add_header_*(). One is byte sequence. Another is null terminated Unicode text.

Add a function bt_obex_string_is_valid() to check if the added string is valid. And add a function bt_obex_unicode_is_valid() dedicated to check Unicode string.

lylezhu2012 committed:
Bluetooth: GOEP: Enable GOEP feature

Add a Kconfig BT_GOEP to control the GOEP feature.

Implement the GOEP protocol and transport, both for GOEP 1.1 and GOEP
2.x.

For GOEP transport, OBEX over RFCOMM, and OBEX over L2CAP are
supported.

For GOEP protocol, `put`, `get`, `abort`, `setpath`, and `action` are
supported. And only one operation can be processed at the same time.
The feature `Reliable Session` is unsupported.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: shell: Add transport commands for GOEP

Add commands for GOEP to test transport features.

Add commands for transport over RFCOMM, including `register-rfcomm`,
`connect-rfcomm`, and `disconnect-rfcomm`.

Add commands for transport over L2CAP, including `register-l2cap`,
`connect-l2cap`, and `disconnect-l2cap`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented:
Another massive PR? :)

@lylezhu2012 I think the L2CAP and RFCOMM improvements could at least be split out into their own PRs. That way I think we can make faster progress with getting functionality merged incrementally.

lylezhu2012 commented:
> Another massive PR? :)

Actually, my PR only implements basic goep/obex operations, but goep and obex define too many concepts.

> @lylezhu2012 I think the L2CAP and RFCOMM improvements could at least be split out into their own PRs. That way I think we can make faster progress with getting functionality merged incrementally.

Well, I will remove all other layers changes from this PR. And make small PR for these changes.

lylezhu2012 commented:
Split out the changes and create other 4 PRs, https://github.com/zephyrproject-rtos/zephyr/pull/83783, https://github.com/zephyrproject-rtos/zephyr/pull/83785, https://github.com/zephyrproject-rtos/zephyr/pull/83786, and https://github.com/zephyrproject-rtos/zephyr/pull/83787.

But the changes is not removed from current PR to avoid the building issue.

kartben commented:
test failure is unrelated and will be fixed with #83907

makeshi commented on include/zephyr/bluetooth/classic/obex.h:
```diff
@@ -0,0 +1,1443 @@
+#include <errno.h>
+#include <stdbool.h>
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_OBEX_H_
```
Generally, the header guard should be placed at the very beginning of the header file, it should be before #include

lylezhu2012 replied:
Updated

makeshi commented on include/zephyr/bluetooth/classic/obex.h:
```diff
@@ -0,0 +1,1443 @@
+
+/** Converts a OBEX response code to string.
+ *
+ * @return The string representation of the response code code.
```
duplicate typo "code"

lylezhu2012 replied:
Updated

makeshi commented on include/zephyr/bluetooth/classic/obex.h:
```diff
@@ -0,0 +1,1443 @@
+};
+
+/** Converts a OBEX response code to string.
+ *
```
No @param doxygen

lylezhu2012 replied:
Updated

lylezhu2012 committed:
Bluetooth: OBEX: Check if the added string is valid

There are types of string can be added by using OBEX adding header
function bt_obex_add_header_*(). One is byte sequence. Another is null
terminated Unicode text.

Add a function bt_obex_string_is_valid() to check if the added string
is valid. And add a function bt_obex_unicode_is_valid() dedicated to
check Unicode string.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on include/zephyr/bluetooth/classic/obex.h:
```diff
@@ -0,0 +1,1443 @@
+/* obex.h - IrDA Oject Exchange Protocol handling */
```
The full name of `obex` is `IrDA Object Exchange Protocol`.

lylezhu2012 committed:
Bluetooth: GOEP: Improve TLV triplet encoding and parsing

Add a structure `struct bt_obex_tlv` to pass/save the one TLV info.

Use a array of `struct bt_obex_tlv` as the set of all TLV-triplets.
Update the following functions to  support the feature, including
`bt_obex_add_header_app_param`, `bt_obex_add_header_auth_challenge`,
and `bt_obex_add_header_auth_rsp`.

Add a function `bt_obex_tlv_parse` to parse the encoded TLV-triplets.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
test: shell: GOEP: Optimize TLV-triplet setting

Add a TLV array `tlvs` to keep the pending TLV data.

Update `app_param`, `auth_challenge`, and `auth_rsp` to set the one
TLV information each time. When the optional argument `last` set,
write the all pending TLVs to the tx buffer.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: GOEP: Support the empty Name header case

According to the OBEX Version 1.5, the Name header could be a empty
string.

Update function `bt_obex_add_header_name` to support the case that the
length of name could be 0.

Update function `bt_obex_get_header_name` to support the case that the
name header is found but the name length is 0.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
Update the PR to support the L2CAP ERET Mode since the L2CAP RET mode is supported now.

lylezhu2012 committed:
Bluetooth: shell: Add OBEX commands for GOEP

Add commands for GOEP to test OBEX features.

Add command `alloc-buf` and `release-buf` to allocate and release TX
buffer.

Add command set `add-header` to add the OBEX headers to allocated TX
buffer.

Add command set `client` to send OBEX client requests with allocated
TX buffer.

Add command set `server` to send OBEX responses with allocated TX
buffer.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
> Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [19 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=zephyrproject-rtos_zephyr&pullRequest=83781&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)

I think this is a negative detective. The arguments  `char *argv[]` should not be a null pointer. And null-terminated has been appended to the end of the string of the arguments  `char *argv[]`. This is ensured by the shell driver.

@jhedberg , Do you have any suggestions? 

And I find the detective results can be deactivated by setting them to "safe" status.

jhedberg commented:
> > Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [19 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=zephyrproject-rtos_zephyr&pullRequest=83781&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)
> 
> I think this is a negative detective. The arguments `char *argv[]` should not be a null pointer. And null-terminated has been appended to the end of the string of the arguments `char *argv[]`. This is ensured by the shell driver.
> 
> @jhedberg , Do you have any suggestions?
> 
> And I find the detective results can be deactivated by setting them to "safe" status.

Looks like those are indeed all false positives likely stemming from the SonarQube not being aware of the relationship and API promise given by `args` and `argv`. I don't know what you mean by "can be deactivated" though since I didn't find any such option. @nashif any help here?

jhedberg commented:
@lylezhu2012 I think we can ignore those failures for now, since they shouldn't block the PR from getting merged.

jhedberg commented on include/zephyr/bluetooth/classic/goep.h:
```diff
@@ -0,0 +1,352 @@
+/**
+ * @brief Generic Object Exchange Profile (GOEP)
+ * @defgroup bt_goep Generic Object Exchange Profile (GOEP)
+ * @since 1.0
```
Isn't this a reference to a Zephyr version?

lylezhu2012 replied:
Updated.

jhedberg commented on include/zephyr/bluetooth/classic/goep.h:
```diff
@@ -0,0 +1,352 @@
+ * @brief Generic Object Exchange Profile (GOEP)
+ * @defgroup bt_goep Generic Object Exchange Profile (GOEP)
+ * @since 1.0
+ * @version 1.0.0
```
This is not correct for a newly introduced API, since starting with `1` implies a stable API. See https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html

lylezhu2012 replied:
Updated.

lylezhu2012 commented on include/zephyr/bluetooth/classic/goep.h:
```diff
@@ -0,0 +1,340 @@
+	 *  true - peer supports GOEP v2.0 or later. The GOEP transport is based on L2CAP.
+	 *  `chan` is used as transport.
+	 */
+	bool _goep_v2;
```
I think `is` is used to check the value of `_goep_v2`. It is not used to modify `_goep_v2`.

lylezhu2012 commented on include/zephyr/bluetooth/classic/obex.h:
```diff
@@ -0,0 +1,1522 @@
+	 *  @param buf Sequence of headers.
+	 */
+	void (*connect)(struct bt_obex *obex, uint8_t rsp_code, uint8_t version, uint16_t mopl,
+			struct net_buf *buf);
```
It will be called whenever the OBEX connect response is received.

lylezhu2012 committed:
Bluetooth: GOEP: Improve the MTU configuration

The MTU of the GOEP should be not less than 255. So, if the transport
is RFCOMM, the CONFIG_BT_BUF_ACL_RX_SIZE should be not less than 264.
It includes,
- 255 bytes for the minimum MTU of GOEP,
- 4 bytes for L2CAP Header,
- 5 bytes for RFCOMM header and FCS.

And if the transport is L2CAP, the CONFIG_BT_BUF_ACL_RX_SIZE should
be not less than 259.
It includes,
- 255 bytes for the minimum MTU of GOEP,
- 4 bytes for L2CAP Header.

Add Kconfig `BT_GOEP_RFCOMM_MTU` to configure the maximum size for
RFCOMM transport. The range of `BT_GOEP_RFCOMM_MTU` is
`[264, BT_RFCOMM_L2CAP_MTU]`. And the GOEP MTU via RFCOMM transport
should be in the range `[255, (BT_GOEP_RFCOMM_MTU-9)]`.

Add Kconfig `BT_GOEP_L2CAP_MTU` to configure the maximum size for
L2CAP transport. The range of `BT_GOEP_L2CAP_MTU` is
`[259, BT_BUF_ACL_RX_SIZE]`. And the GOEP MTU via L2CAP transport
should be in the range `[255, (BT_GOEP_L2CAP_MTU-4)]`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on include/zephyr/bluetooth/classic/goep.h:
```diff
@@ -0,0 +1,340 @@
+	 *  If this callback is provided it will be called whenever the GOEP transport is
+	 *  disconnected, including when a connection gets rejected.
+	 *
+	 *  @param goep The GOEP object that has been Disconnected.
```
Updated.

lylezhu2012 commented on include/zephyr/bluetooth/classic/obex.h:
```diff
@@ -0,0 +1,1522 @@
+/** @brief Add Header: oBEX Object class of object.
+ *
+ *  @param buf Buffer needs to be sent.
+ *  @param len Length of oject class.
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/obex.c:
```diff
@@ -0,0 +1,3589 @@
+	uint16_t index = 0;
+	uint16_t code;
+
+	if ((len < 2) || (len % 2)) {
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/goep.c:
```diff
@@ -0,0 +1,1646 @@
+	goep->conn = NULL;
+}
+
+void goep_transport_connected(struct bt_conn *conn, struct bt_goep *goep)
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/goep.c:
```diff
@@ -0,0 +1,701 @@
+	return bt_goep_create_pdu(goep, pool);
+}
+
+static int geop_l2cap_disconnect(struct bt_obex *obex)
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -508,6 +508,48 @@ config BT_COD
+	bool "Bluetooth GOEP Profile [EXPERIMENTAL]"
+	select BT_RFCOMM
+	select EXPERIMENTAL
+	help
```
Updated.

MarkWangChinese commented:
Hi @lylezhu2012 After reviewing the interface of obex and geop, I have two thoughts:
1. Hide the goep transport interface, only export the obex interface, application only uses the obex interface. How to do it: Expand the `struct bt_obex_transport_ops` to provide all the transport establish/distory interface to obex. Then it is more easy-to-use by application.
2. From the spec, obex is a client/server request-response paradigm. In one obex connection, the obex role can only be client or server. So I think obex interfaces can be splitted as client interface and server interface. Then it is clear and easy-to-use by application.

Since the obex/geop interfaces should not be used from application. Other profiles (like: pbap and map) use obex/geop and export interfaces to application, I think the current implementation is OK to me too.

lylezhu2012 commented on subsys/bluetooth/host/classic/goep.c:
```diff
@@ -0,0 +1,701 @@
+#define GOEP_MIN_MTU BT_OBEX_MIN_MTU
+
+/* RFCOMM Server list */
+static sys_slist_t goep_rfcomm_server;
```
Updated.
